Assets/Scripts/EOSTransferTask.cs

```diff
@@ -0,0 +1,305 @@
+        /// </summary>
+        public Result? ResultCode;
+
+        /// <summary>
+        /// Percentage indicator of the work's completeness.
+        /// Either read from this value or subscribe to <see cref="OnProgressUpdated"/> to receive updates on progress amounts.
+        /// </summary>
+        public float Progress;
+
+        /// <summary>
+        /// Inner backing for the total size of the transfer.
+        /// </summary>
+        private uint transferSize = 0;
+
+        /// <summary>
+        /// Indexes the current byte count that has been processed.
+        /// </summary>
+        private uint currentIndex = 0;
+
+        /// <summary>
+        /// The total size of the transfer operation in bytes.
+        /// </summary>
+        public uint TotalSize
+        {
+            get
+            {
+                return transferSize;
+            }
+            private set
+            {
+                transferSize = value;
+
+                if (transferSize > FILE_MAX_SIZE_BYTES)
+                {
+                    Debug.LogError("[EOS SDK] Player data storage: data transfer size exceeds max file size.");
+                    transferSize = FILE_MAX_SIZE_BYTES;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructor for download scenarios.
+        /// </summary>
+        /// <param name="innerTask">A task completion source that can be used to indicate when the task is completed.</param>
+        public EOSTransferTask(TaskCompletionSource<byte[]> innerTask)
+        {
+            InnerTaskCompletionSource = innerTask;
+        }
+
+        /// <summary>
+        /// Constructor for upload scenarios.
+        /// </summary>
+        /// <param name="innerTask">A task completion source that can be used to indicate when the task is completed.</param>
+        /// <param name="data">The data to upload, in byte[] form.</param>
+        public EOSTransferTask(TaskCompletionSource<byte[]> innerTask, byte[] data)
+        {
+            InnerTaskCompletionSource = innerTask;
+            Data = data;
+            TotalSize = (uint)data.Length;
+        }
+
+        /// <summary>
+        /// Constructor for immediate completion scenarios.
+        /// </summary>
+        /// <param name="resultCode">The result code to set.</param>
+        /// <param name="failureException">An optional exception that explains the result.</param>
+        public EOSTransferTask(Result resultCode, Exception failureException = null)
+        {
+            InnerTaskCompletionSource = new TaskCompletionSource<byte[]>();
+            ResultCode = resultCode;
+
+            if (failureException != null)
+            {
+                InnerTaskCompletionSource.SetException(failureException);
+            }
+            else
+            {
+                InnerTaskCompletionSource.SetCanceled();
+            }
+        }
```

Comment:
It is unclear to me when this would be utilized - it seems like if you know that the failure is immediate there is no need to generate the task. Is there additional context that I am missing?